### PR TITLE
Remote notification completion handler

### PR DESCRIPTION
**Pull request by JAStanton**
**Motivation**
To allow handling of remote notifications using the preferred delegate [application:didReceiveRemoteNotification:fetchCompletionHandler](https://developer.apple.com/library/ios/documentation/UIKit/Reference/UIApplicationDelegate_Protocol/#//apple_ref/occ/intfm/UIApplicationDelegate/application:didReceiveRemoteNotification:fetchCompletionHandler:).

React native right now is setup to use [application:didReceiveRemoteNotification](https://developer.apple.com/library/ios/documentation/UIKit/Reference/UIApplicationDelegate_Protocol/#//apple_ref/occ/intfm/UIApplicationDelegate/application:didReceiveRemoteNotification:) which doesn't give the engineer the option to call back to the app via a completion handler and inform the app that it is done with handling the push notification.

The primary reason to use the `fetchCompletionHandler` would be to wake up the app to fetch updates so the next time the user opens the app it will be up to date.

A new method will be exposed: `PushNotificationIOS#finish` which can be called for remote notifications.

**Test Plan**
The new flow would be something like this:

Set up the app to receive a notification and then register for `_onNotification` listener
```js
PushNotification.addEventListener('notification', this._onNotification);
```

Trigger a remote notification and then handle it like so:
```js
_onNotification(notification) {
  // handle notification, and then ...
  notification.finish(PushNotification.FetchResult.NewData);
}
```

### CODE REVIEW

**Comment by facebook-github-bot-5** [comment]
By analyzing the blame information on this pull request, we identified @nicklockwood and @slycoder to be potential reviewers.

**Comment by mkonicek** [comment]
cc @javache

**Comment by fgaleano** [comment]
@JAStanton Will this change allow for updating the icon badge when a notification comes in and the app is in the background? I've noticed my app badge only increments upon opening the app but not when the remote notification comes in and I'm on the home screen or another app. Thanks.

**Comment by JAStanton** [comment]
@fgaleano that's not something I set out to fix. But I don't recall it being an issue before and using my app now that relies on this PR it definitely works. Sorry for the wishy washy answer it's just hard for me to test the old behavior now that i'm so far down this path :P

**Comment by facebook-github-bot-7** [comment]
It's been a while since the last commit was reviewed and the labels show this pull request needs review. Based on the blame information for the files in this pull request we identified @nicklockwood as a potential reviewer. Could you take a look please or cc someone with more context?

File: Libraries/PushNotificationIOS/PushNotificationIOS.js [hunk 1]
```diff
@@ -172,13 +193,15 @@ class PushNotificationIOS {
       listener =  PushNotificationEmitter.addListener(
         DEVICE_NOTIF_EVENT,
         (notifData) => {
+          notifData.isRemote = true;
```
Comment (javache) [inline Libraries/PushNotificationIOS/PushNotificationIOS.js]: remote already got added in another pull-request. Could you rebase this?

Comment (JAStanton) [inline Libraries/PushNotificationIOS/PushNotificationIOS.js]: Hmmm well looking at Github's master I see this: https://github.com/facebook/react-native/blob/master/Libraries/PushNotificationIOS/PushNotificationIOS.js#L175-L177 which does NOT have isRemote. Maybe you're seeing something I'm not? Possibly something that's internal to Facebook members only? I don't know. I just updated with master and this is the diff though.

**Comment by JAStanton** [comment]
Any updates on this? @javache

File: Libraries/PushNotificationIOS/PushNotificationIOS.js (hunk 1 above)
Comment (javache) [inline Libraries/PushNotificationIOS/PushNotificationIOS.js]: This is now set on the native side: https://github.com/facebook/react-native/blob/master/Libraries/PushNotificationIOS/RCTPushNotificationManager.m#L162

Comment (JAStanton) [inline Libraries/PushNotificationIOS/PushNotificationIOS.js]: ah brilliant! ok I'll update my PR to rely on it then. Thanks

**Comment by facebook-github-bot-3** [comment]
@JAStanton updated the pull request.

**Comment by facebook-github-bot-6** [comment]
@JAStanton updated the pull request.

**Comment by JAStanton** [comment]
Alright @javache I updated this PR with your comments

**Comment by JAStanton** [comment]
Any movement?

**Comment by JAStanton** [comment]
@javache any movement?

File: Libraries/PushNotificationIOS/RCTPushNotificationManager.m [hunk 2]
```diff
@@ -31,6 +31,11 @@
 
 NSString *const RCTErrorUnableToRequestPermissions = @"E_UNABLE_TO_REQUEST_PERMISSIONS";
 
+@interface RCTPushNotificationManager ()
+@property (nonatomic, copy) RCTPromiseResolveBlock requestPermissionsResolveBlock;
+@property (strong, nonatomic) NSMutableDictionary* remoteNotificationCallbacks;
```
Comment (javache) [inline Libraries/PushNotificationIOS/RCTPushNotificationManager.m]: nit: our objc style is ` @property (nonatomic, strong) NSMutableDictionary *remoteNotificationCallbacks`

File: Libraries/PushNotificationIOS/RCTPushNotificationManager.h [hunk 3]
```diff
@@ -11,9 +11,12 @@
 
 @interface RCTPushNotificationManager : RCTEventEmitter
 
+typedef void (^RemoteNotificationCallback)(UIBackgroundFetchResult result);
```
Comment (javache) [inline Libraries/PushNotificationIOS/RCTPushNotificationManager.h]: These typenames are global, can you prefix it with RCT?

File: Libraries/PushNotificationIOS/RCTPushNotificationManager.m [hunk 4]
```diff
@@ -158,8 +179,20 @@ - (void)handleLocalNotificationReceived:(NSNotification *)notification
...
+  remoteNotification[@"remote"] = @YES;
+  if (completionHandler) {
+    if (self.remoteNotificationCallbacks == nil) {
+      // Lazy initialization
+      self.remoteNotificationCallbacks = [NSMutableDictionary dictionary];
+    }
+    self.remoteNotificationCallbacks[notificationId] = completionHandler;
+  }
+
+  [self sendEventWithName:@"remoteNotificationReceived" body:notification.userInfo];
```
Comment (javache) [inline Libraries/PushNotificationIOS/RCTPushNotificationManager.m]: Why are you sending the event twice?

File: Libraries/PushNotificationIOS/RCTPushNotificationManager.m [hunk 5]
```diff
@@ -158,8 +179,20 @@ - (void)handleLocalNotificationReceived:(NSNotification *)notification
...
 {
-  NSMutableDictionary *userInfo = [notification.userInfo mutableCopy];
-  userInfo[@"remote"] = @YES;
+  NSMutableDictionary *remoteNotification = [NSMutableDictionary dictionaryWithDictionary:notification.userInfo[@"notification"]];
+  RemoteNotificationCallback completionHandler = notification.userInfo[@"completionHandler"];
+  NSString* notificationId = [[NSUUID UUID] UUIDString];
+  remoteNotification[@"notificationId"] = notificationId;
+  remoteNotification[@"remote"] = @YES;
+  if (completionHandler) {
+    if (self.remoteNotificationCallbacks == nil) {
```
Comment (javache) [inline Libraries/PushNotificationIOS/RCTPushNotificationManager.m]: `if (!self.remoteNotificationCallbacks)`

File: Libraries/PushNotificationIOS/RCTPushNotificationManager.m [hunk 6]
```diff
@@ -185,6 +218,20 @@ - (void)handleRegisterUserNotificationSettings:(NSNotification *)notification
   _requestPermissionsResolveBlock = nil;
 }
 
+RCT_EXPORT_METHOD(onFinishRemoteNotification:(NSString*)notificationId fetchResult:(UIBackgroundFetchResult)result) {
+  RemoteNotificationCallback completionHandler = self.remoteNotificationCallbacks[notificationId];
+  if (completionHandler == nil) {
+    @throw [NSException
```
Comment (javache) [inline Libraries/PushNotificationIOS/RCTPushNotificationManager.m]: This doesn't seem like it should be a fatal error. Instead use RCTLogError (it will show a redbox too)

File: Libraries/PushNotificationIOS/RCTPushNotificationManager.m [hunk 7]
```diff
@@ -185,6 +218,20 @@ - (void)handleRegisterUserNotificationSettings:(NSNotification *)notification
   _requestPermissionsResolveBlock = nil;
 }
 
+RCT_EXPORT_METHOD(onFinishRemoteNotification:(NSString*)notificationId fetchResult:(UIBackgroundFetchResult)result) {
+  RemoteNotificationCallback completionHandler = self.remoteNotificationCallbacks[notificationId];
+  if (completionHandler == nil) {
```
Comment (javache) [inline Libraries/PushNotificationIOS/RCTPushNotificationManager.m]: `if (!completionHandler)`

**Comment by javache** [comment]
Looks good, just a couple of style nits

File: Libraries/PushNotificationIOS/RCTPushNotificationManager.m [hunk 8]
```diff
@@ -31,6 +31,11 @@
 
 NSString *const RCTErrorUnableToRequestPermissions = @"E_UNABLE_TO_REQUEST_PERMISSIONS";
 
+@interface RCTPushNotificationManager ()
+@property (nonatomic, copy) RCTPromiseResolveBlock requestPermissionsResolveBlock;
```
Comment (javache) [inline Libraries/PushNotificationIOS/RCTPushNotificationManager.m]: Why did you add this?

File: Libraries/PushNotificationIOS/RCTPushNotificationManager.m (hunk 4 above)
Comment (JAStanton) [inline Libraries/PushNotificationIOS/RCTPushNotificationManager.m]: Bad merge

File: Libraries/PushNotificationIOS/RCTPushNotificationManager.m (hunk 8 above)
Comment (JAStanton) [inline Libraries/PushNotificationIOS/RCTPushNotificationManager.m]: Bad merge

**Comment by facebook-github-bot-7** [comment]
@JAStanton updated the pull request - [view changes](https://github.com/facebook/react-native/pull/8040/files/b1266eebc0b1e411132475776263ddadd5b4bd93..5095708f0c2f835a9ad99b427787e6fe36727b25)

**Comment by facebook-github-bot-2** [comment]
@JAStanton updated the pull request - [view changes](https://github.com/facebook/react-native/pull/8040/files/5095708f0c2f835a9ad99b427787e6fe36727b25..c59a013f7a8a879761789f3f5d0d005b42733ad6)

**Comment by facebook-github-bot-9** [comment]
@JAStanton updated the pull request - [view changes](https://github.com/facebook/react-native/pull/8040/files/c59a013f7a8a879761789f3f5d0d005b42733ad6..541437336d97f3b909d1159281a1beed87801df1)

File: Libraries/PushNotificationIOS/RCTPushNotificationManager.m (hunk 8 above)
Comment (JAStanton) [inline Libraries/PushNotificationIOS/RCTPushNotificationManager.m]: Whoops :) Bad merge.

**Comment by facebook-github-bot-5** [comment]
@JAStanton updated the pull request - [view changes](https://github.com/facebook/react-native/pull/8040/files/541437336d97f3b909d1159281a1beed87801df1..1d303c05c4e2f5e6b664d6ce4441df164330acd7)

**Comment by facebook-github-bot-3** [comment]
@JAStanton updated the pull request - [view changes](https://github.com/facebook/react-native/pull/8040/files/1d303c05c4e2f5e6b664d6ce4441df164330acd7..734ef9758defdcea21f9035b887bcf1f58f2b2e3)

**Comment by facebook-github-bot-5** [comment]
@JAStanton updated the pull request - [view changes](https://github.com/facebook/react-native/pull/8040/files/1d303c05c4e2f5e6b664d6ce4441df164330acd7..734ef9758defdcea21f9035b887bcf1f58f2b2e3)

**Comment by facebook-github-bot-8** [comment]
@JAStanton updated the pull request - [view changes](https://github.com/facebook/react-native/pull/8040/files/734ef9758defdcea21f9035b887bcf1f58f2b2e3..db74ec897413d6c9f227c1257464f01f46b0e258)

**Comment by JAStanton** [comment]
@javache I'm not a Flow user, I attempted to fix https://circleci.com/gh/facebook/react-native/10648 (https://github.com/facebook/react-native/pull/8040/commits/734ef9758defdcea21f9035b887bcf1f58f2b2e3) with https://github.com/facebook/react-native/pull/8040/commits/db74ec897413d6c9f227c1257464f01f46b0e258 but I ended up getting linter errors. I know it must be simple, but how do I put a typed static that wont blow up the website documentation generator?

**Comment by javache** [comment]
@bestander / @hramos Can you help here?

**Comment by bestander** [comment]
Yeah, the website EcmaScript parser is quite outdated and we start stepping on it more and more often.
cc @lacker 

Flow issue should be possible to fix by declaring a static on the class.
In the PR you are assigning static FetchResult field to a class that does not have it declared.

**Comment by JAStanton** [comment]
@bestander 

> Flow issue should be possible to fix by declaring a static on the class

I originally declared a static on the class: https://github.com/facebook/react-native/pull/8040/commits/734ef9758defdcea21f9035b887bcf1f58f2b2e3 but I got this error: https://circleci.com/gh/facebook/react-native/10648

Thats when I changed up the syntax a bit https://github.com/facebook/react-native/pull/8040/commits/db74ec897413d6c9f227c1257464f01f46b0e258 but i ended up getting the current failures https://circleci.com/gh/facebook/react-native/10653

**Comment by bestander** [comment]
There is internal task raised to update the ES Parser 12652958.
Meanwhile, @JAStanton, could you experiment with the static FetchResult?
Maybe if you split definition and assignment it will pass

**Comment by lacker** [comment]
The website doc generator is updated so it should work with statics on the class now. Let me know if that is not the case

**Comment by lacker** [comment]
Travis looks OK, it looks like that run was canceled rather than failing. So I think this should be good to go - there's a merge conflict though. Can you fix that conflict and then I will ship this?

**Comment by JAStanton** [comment]
@lacker alright updated, now lets wait a few hours to see if CI runs successfully :P

**Comment by lacker** [comment]
Eh, Travis is failing but it looks unrelated - there is some logging test that's failing on tvOS. @bestander what would you normally do in this situation?

**Comment by bestander** [comment]
Tests look green now

**Comment by JAStanton** [comment]
@lacker @bestander :shipit: ?

**Comment by bestander** [comment]
Import failed:

```
Libraries/FBReactKit/js/react-native-github/Libraries/PushNotificationIOS/RCTPushNotificationManager.m:261:17: error: format string is not a string literal (potentially insecure) [-Werror,-Wformat-security]
    RCTLogError([NSString stringWithFormat:
    ~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~
Libraries/FBReactKit/js/react-native-github/React/Base/RCTLog.h:35:52: note: expanded from macro 'RCTLogError'
#define RCTLogError(...) _RCTLog(RCTLogLevelError, __VA_ARGS__)
                         ~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~
Libraries/FBReactKit/js/react-native-github/React/Base/RCTLog.h:127:74: note: expanded from macro '_RCTLog'
#define _RCTLog(lvl, ...) _RCTLogNativeInternal(lvl, __FILE__, __LINE__, __VA_ARGS__);
                                                                         ^~~~~~~~~~~
Libraries/FBReactKit/js/react-native-github/Libraries/PushNotificationIOS/RCTPushNotificationManager.m:261:17: note: treat the string as an argument to avoid this
    RCTLogError([NSString stringWithFormat:
                ^
                @"%@", 
Libraries/FBReactKit/js/react-native-github/React/Base/RCTLog.h:35:52: note: expanded from macro 'RCTLogError'
#define RCTLogError(...) _RCTLog(RCTLogLevelError, __VA_ARGS__)
                                                   ^
Libraries/FBReactKit/js/react-native-github/React/Base/RCTLog.h:127:74: note: expanded from macro '_RCTLog'
#define _RCTLog(lvl, ...) _RCTLogNativeInternal(lvl, __FILE__, __LINE__, __VA_ARGS__);
                                                                         ^
1 error generated.
```

**Comment by JAStanton** [comment]
And so it continues.

File: Libraries/PushNotificationIOS/RCTPushNotificationManager.m [hunk 9]
```diff
@@ -224,6 +255,18 @@ - (void)handleRegisterUserNotificationSettings:(NSNotification *)notification
   _requestPermissionsResolveBlock = nil;
 }
 
+RCT_EXPORT_METHOD(onFinishRemoteNotification:(NSString*)notificationId fetchResult:(UIBackgroundFetchResult)result) {
+  RCTRemoteNotificationCallback completionHandler = self.remoteNotificationCallbacks[notificationId];
+  if (!completionHandler) {
+    RCTLogError([NSString stringWithFormat:
```
Comment (lacker) [inline Libraries/PushNotificationIOS/RCTPushNotificationManager.m]: I think this is the problem - you don't need a `stringWithFormat` inside `RCTLogError`, because `RCTLogError` already takes a first argument with `@`'s in it and does this interpolation. Could you fix this?

Comment (JAStanton) [inline Libraries/PushNotificationIOS/RCTPushNotificationManager.m]: done

**Comment by JAStanton** [comment]
@lacker fingers crossed.

**Comment by lacker** [comment]
That Travis failure looks like nonsense. I'm just gonna shipit.

**Comment by lacker** [comment]
@facebook-github-bot shipit

**Comment by JAStanton** [comment]
139 days later!
image

**Comment by npomfret** [comment]
Thank you all so much!! You wouldn't believe the nonsense I've been trying to work around this.

**Comment by JAStanton** [comment]
:) glad to hear it!

**Comment by npomfret** [comment]
I've just tried this on on 0.38.rc and i've kind of fallen at the 1st hurdle... My AppDelegate has this:

```
-(void)application:(UIApplication *)application didReceiveRemoteNotification:(NSDictionary *)notification fetchCompletionHandler:(void (^)(UIBackgroundFetchResult))completionHandler { 

NSLog(@"push-notification with content body => %@", notification);

  [RCTPushNotificationManager didReceiveRemoteNotification:notification fetchCompletionHandler:completionHandler];
}
```

My I'm just seeing the red error screen with 

`ExceptionsManager.js:71 RCTJSONStringify() encountered the following error: Invalid type in JSON write (__NSMallocBlock__)`

And the notification that gets delivered to my JS code is empty (all the fields have _undefined_ values).

I can see the notification being delivered the the AppDelegate correctly.

Any idea what I might be doing wrong?

**Comment by npomfret** [comment]
I think I found the problem - the wrong object is being sent

https://github.com/facebook/react-native/pull/10860

**Comment by npomfret** [comment]
Hi.  Still having issues getting this to work end-to-end.  I've patched my code with #10860 and I'm calling `[RCTPushNotificationManager didReceiveRemoteNotification:notification fetchCompletionHandler:completionHandler];` but I see the following in the iOS logs:

`Warning: Application delegate received call to -application:didReceiveRemoteNotification:fetchCompletionHandler: but the completion handler was never called`

I've implemented my JS to call the new `finish` method, but it's not being given a chance to run - the log message above happens immediately.

I'm not exactly sure but i think the problem might be that `didReceiveRemoteNotification` needs to either call the completion handler in-process (which isn't possible here because of the async call to the JS code), or it needs to start a background task in which we call the completion handler.  The background tasks keeps the app from being suspended for up to 30s.  When the JS code calls `finish` the task can call the completion handler and then end.

https://forums.developer.apple.com/message/96731#96731
